Private/rotem/new domains

rotelimelech opened the pull request:
Adding new domains to allow us to search with BOINC for more results.
This includes:
1. zeta5
2. zeta7 
3. an improve Cartesian domain 
4. catalan
Also, allowing FR enumerator to search for expressions with more then one constant

ShaharGottlieb commented on ramanujan/enumerators/FREnumerator.py:
```diff
@@ -111,26 +111,44 @@ def _improve_results_precision(self, intermediate_results, verbose=True):
-        const = self.constants_generator[0]() # using only one constant for now.
+        # The expression PSLQ tries to find is
+        # (a + b * const) / (c + d * const) = val
+        # => a + b*const -c*val -c*d*val = 0
```
probably meant
a+b * const - c * val - d * val * const =0
?

ShaharGottlieb commented on ramanujan/enumerators/FREnumerator.py:
```diff
@@ -111,26 +111,44 @@ def _improve_results_precision(self, intermediate_results, verbose=True):
+                mpf_val = mpmath.mpf(val)
+                denom_items = [-mpf_val * c for c in numer_items]
                 pslq_res = mpmath.pslq(
-                    [1, const, const**2, -mpf_val, -const * mpf_val, -(const**2) * mpf_val],
```
i see that you left out the const**2 term.
please make sure that you passed all unit tests now.

if you didn't need to modify any of the tests, then it is a sign you are not testing enough :)

rotelimelech replied:
I've modified the test to use a smaller list when matching results, and they all pass :)

However there is no testing for more then one constant. I'll add some test for that

ShaharGottlieb commented on ramanujan/poly_domains/CartesianProductPolyDomain.py:
```diff
@@ -4,17 +4,31 @@
-        it is a redundant run we can skip. Based on an_leading_coef_positive we will try to detect those cases and skip
-        them
+        a_deg - an's polynomial degree
+        a_coef_range - The coef range itered for every coef in an
```
coefficient. 

please refrain from abbreviating in documentation

ShaharGottlieb commented on ramanujan/poly_domains/CartesianProductPolyDomain.py:
```diff
@@ -4,17 +4,31 @@
+        b_deg - bn's polynomial degree
+        b_coef_range - The coef range itered for every coef in bn
+        an_leading_coef_positive - A GCF can be inflated by multiplying an*c and bn*c^2. The inflation will converge to
+            the same value (up to multiplying by a factor). By forcing an_leading coef to be positive we can remove
```
.

ShaharGottlieb commented on ramanujan/poly_domains/CartesianProductPolyDomain.py:
```diff
@@ -52,43 +68,116 @@ def domain_size_by_var_ranges(var_ranges):
+    def filter_gcfs(self, an_coefs, bn_coefs):
+        """
+        Some GCFs will not converge, and some are duplicates of other GCFs
+        This function filter those cases out
```
filters these*

ShaharGottlieb commented on ramanujan/poly_domains/ExplicitCartesianProductPolyDomain.py:
```diff
@@ -0,0 +1,21 @@
+
+        # This domain's degree is not constant and needs to be calculated 
+        self.a_deg = len(self.a_coef_range) - 1
+        self.b_deg = len(self.b_coef_range) - 1
```
no newline at end of file

ShaharGottlieb reviewed (commented):
overall looks ok! general comments:
1 - in all documentation, use "coefficient" instead of "coef"
2 - since we added functionality, i think we should add test cases as well. especially node when we have actual users.. it is a lot more scarier to have a bug in production. also it is good for the next changes, to see that nothing "breaks".
3 - how do you prepare to monitor bugs from users?
this is something to think about... maybe we need a "send report" option? log dump? instructions leading to issues tap in the repo?

rotelimelech commented:
Fixed all comments,
I've also added a test that uses more then one constant, this will also check if the feature of creating a PSLQ vector works well.
